docs/source/architecture.md

```diff
@@ -95,16 +95,41 @@ so less restrictions apply to this than to `get_build_scripts`.
 
 At the end of the assemble step, the docker image is ready to be used in various ways!
 
-## Push
+### Push
 
 Optionally, repo2docker can **push** a built image to a [docker registry](https://docs.docker.com/registry/).
 This is done as a convenience only (since you can do the same with a `docker push` after using repo2docker
 only to build), and implemented in `Repo2Docker.push` method. It is only activated if using the
 `--push` commandline flag.
 
-## Run
+### Run
 
 Optionally, repo2docker can **run** the built image and allow the user to access the Jupyter Notebook
 running inside by default. This is also done as a convenience only (since you can do the same with `docker run`
 after using repo2docker only to build), and implemented in `Repo2Docker.run`. It is activated by default
-unless the `--no-run` commandline flag is passed.
\ No newline at end of file
+unless the `--no-run` commandline flag is passed.
+
+## ContentProviders
+
+ContentProviders provide a way for `repo2docker` to know how to find and
+retrieve a repository. They follow a similar pattern as the BuildPacks
+described above. When `repo2docker` is called, its main argument will be
+a path to a repository. This might be a local path or a URL. Upon being called,
+`repo2docker` will loop through all ContentProviders and perform the following
+commands:
+
+* Run the `detect(` method on the repository path given to `repo2docker`. This
```

Comment:
```suggestion
* Run the `detect()` method on the repository path given to `repo2docker`. This
```
I've not seen the `detect(` notation before and find it a bit weird, it feels like writing a sentence with no punctuation at the end 

Somehow unfinished :)